### PR TITLE
Simplify merkledag package interface to ease IPLD inclusion

### PR DESCRIPTION
**Pull request by mildred**
- Remove usage of merkledag.Link.Node pointer outside of merkledag
- Make Node.Unmarshal() private: it is not used elsewhere
- Rename Decoded(= into DecodeProtobuf() to clearly give intent about the byte wire format

Obsoletes #1916 and #1902.

This prepares for inclusion of IPLD where the Node pointer might not be publicly accessible of even be there. We should probably move away from using the struct attributes and use functions to access them instead. This will make it easier to plug in IPLD in place of merkledag.

### CODE REVIEW

**Comment by mildred** [comment]
- AppVeyor fails with `install deps: failed to fetch dependencies`
- Travis also fails with dependency error, except on Windows where I don't understand what is the error

Is there a way to manually trigger the build so I could test what does wrong incrementally?

**Comment by chriscool** [comment]
> Travis also fails with dependency error, except on Windows where I don't understand what is the error

You mean on MacOS as Travis runs on MacOS and Linux.
Anyway all the errors on Travis and Appveyor seem to happen when fetching dependencies.

**Comment by whyrusleeping** [comment]
the appveyor one is now better, its moved on to failing on sharness tests. The travis one is weird... its hanging while contacting the gateways for some reason.

That aside, I think now is a good opportunity to, if not remove, at least unexport the `Node` field, i'm in favor of removing it if youre feeling adventurous.

File: path/resolver.go [hunk 1]
```diff
@@ -111,37 +111,33 @@ func (s *Resolver) ResolveLinks(ctx context.Context, ndd *merkledag.Node, names
...
-		if next == "" {
+		if nlink == nil || len(nlink.Hash) == 0 {
 			n, _ := nd.Multihash()
 			return result, ErrNoLink{Name: name, Node: n}
 		}
 
-		if nlink.Node == nil {
-			// fetch object for link and assign to nd
-			ctx, cancel := context.WithTimeout(ctx, time.Minute)
+		if nlink.GetCachedNode() == nil {
```
Comment (whyrusleeping) [inline path/resolver.go]: this doesnt seem right now... We should probably just drop this timeout logic altogether

File: path/resolver.go [hunk 2]
```diff
@@ -111,37 +111,33 @@ func (s *Resolver) ResolveLinks(ctx context.Context, ndd *merkledag.Node, names
...
-				return append(result, nd), err
-			}
-			nlink.Node = nd
-		} else {
-			nd = nlink.Node
 		}
 
-		result = append(result, nlink.Node)
+		var err error
+		nd, err = nlink.GetNodeAndCache(ctx, s.DAG)
```
Comment (whyrusleeping) [inline path/resolver.go]: most of the logic in this loop could be replaced with:
```
nd, err := nd.GetLinkedNode(ctx, s.DAG, name)
...
```

File: path/resolver.go (hunk 1 above)
Comment (mildred) [inline path/resolver.go]: I tried not to change the code logic at all, but you are right, this could be simplified.

File: path/resolver.go (hunk 2 above)
Comment (mildred) [inline path/resolver.go]: The `GetLinkedNode` creates an unnecessary copy.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: d048b42ec7be8968d9e09b707689359b75ec47fb
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: c39d5ae32fdbd9d8c28b7c3f4debc7c8a37740fd
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by mildred** [comment]
Made the Node field private, and simplified path/resolver.go as you suggested.

**Comment by mildred** [comment]
@whyrusleeping I'm removing AddRecursive and RemoveRecursive so I can finally remove the `Node` field as you suggested. I still have a commit to push and this field is gone.

**Comment by mildred** [comment]
These past weeks, I have been thinking on how to include IPLD in IPFS, and my last idea is to transform the Get() methods on DAGService and NodeGetter to GetPB() methods. And add a Get() method that returns an IPLDNode instead of a protocol buffer Node. Then gradually phase out the GetPB() method in favor of the Get() method.

See https://github.com/mildred/go-ipfs/compare/ipld...mildred:ipld-inclusion

**Comment by mildred** [comment]
Tests are failing since commit "merkledag: Remove cached Node.node …".

**Comment by whyrusleeping** [comment]
I don't know how onboard i am with the `GetPB` change right now, i think we should go ahead and get the earlier cleanup work merged first, and then discuss that change in another PR

**Comment by mildred** [comment]
@whyrusleeping makes sense. I'll leave this commit out for the moment (in my next branch update).

Also, I'd need to discuss how integration of IPLD could be done. Is there an issue for that ? Also, I'll need some review on ipfs/go-ipld#19

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -453,16 +453,20 @@ func (i *gatewayHandler) deleteHandler(w http.ResponseWriter, r *http.Request) {
...
 		if err != nil {
 			webError(w, "Could not update node links", err, http.StatusInternalServerError)
 			return
 		}
 	}
 
-	if err := i.node.DAG.AddRecursive(newnode); err != nil {
-		webError(w, "Could not add recursively new node", err, http.StatusInternalServerError)
+	if _, err := i.node.DAG.Add(newnode); err != nil {
+		webError(w, "Could not add recursively root node", err, http.StatusInternalServerError)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: should probably remove the word recursively here

File: merkledag/traverse/traverse_test.go [hunk 4]
```diff
@@ -341,57 +352,75 @@ func testWalkOutputs(t *testing.T, root *mdag.Node, opts Options, expect []byte)
...
 
-func addChild(t *testing.T, a *mdag.Node, name string) *mdag.Node {
-	c := &mdag.Node{Data: []byte(string(a.Data) + "/" + name)}
-	addLink(t, a, c)
-	return c
+func child(t *testing.T, ds mdag.DAGService, a *mdag.Node, name string) *mdag.Node {
+	return &mdag.Node{Data: []byte(string(a.Data) + "/" + name)}
+}
+
+func newDAGService() mdag.DAGService {
```
Comment (whyrusleeping) [inline merkledag/traverse/traverse_test.go]: you can get this from `merkledag/test.Mock()`

**Comment by whyrusleeping** [comment]
one minor comment and then LGTM

**Comment by mildred** [comment]
Corrected.

**Comment by whyrusleeping** [comment]
Thanks!

**Pull request merged**